File: Elixir/lib/problem063.ex
defmodule Problem063 do
  def can_form_word?(matrix, word) do
    candidates = horizontal_words(matrix) ++ vertical_words(matrix)
    Enum.any?(candidates, &(&1 == String.upcase(word)))
  end

  def horizontal_words(matrix) do
    # Tuples can't be used inside the Enum functions since they're not Enumerable
    Enum.map(matrix, fn x -> Tuple.to_list(x) |> Enum.join end)
  end

  def vertical_words(matrix, index \\ 0, acc \\ []) do
    cond do
      index == length(matrix) ->
        acc
      true ->
        current_word = Enum.map(matrix, fn x -> elem(x, index) end) |> Enum.join
        vertical_words(matrix, index + 1, [current_word | acc])
    end
  end
end

File: Ruby/problem063.rb
class Problem063
  class << self
    def can_form_word?(matrix, word)
      candidates = []
      matrix.each_with_index do |row, index|
        # Get the word in the line
        candidates << row.join
        # Find words vertically
        candidates << (matrix.map { |r| r[index] }).join
      end
      candidates.any? { |item| item == word.upcase }
    end
  end
end
